### PR TITLE
Add a failing test for Unicode conversion

### PR DESCRIPTION
**Pull request by pkese**
### What does this implement/fix? Explain your changes.

A failing test for #1466

### Does this close any currently open issues?

No, just analyzes failure case.

### Any other comments?

See #1466

### Checklist

Check all those that are applicable and complete.

-   [x] Make sure to include one or more tests for your change
-   [ ] If an enhancement PR, please create docs and at best an example
-   [x] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [x] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/pythonnet/pythonnet?pullRequest=1467) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/pythonnet/pythonnet?pullRequest=1467) before we can accept your contribution.<br/><br/>:x: pkese [sign now](https://cla.dotnetfoundation.org/pythonnet/pythonnet?pullRequest=1467)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/pythonnet/pythonnet?pullRequest=1467) it.</sub>

File: src/runtime/runtime.cs [hunk 1]
```diff
@@ -1527,7 +1527,12 @@ internal static bool PyString_Check(IntPtr ob)
         internal static IntPtr PyString_FromString(string value)
         {
             fixed(char* ptr = value)
-                return PyUnicode_FromKindAndData(2, (IntPtr)ptr, value.Length);
+                return Delegates.PyUnicode_DecodeUTF16(
+                    (IntPtr)ptr,
+                    new IntPtr(value.Length * sizeof(Char)),
```
Comment (lostmsu) [inline src/runtime/runtime.cs:1532]: We're in C# 9 now, so we can use `nint` instead of `IntPtr` for `ssize_t`

File: src/runtime/runtime.cs [hunk 2]
```diff
@@ -1527,7 +1527,12 @@ internal static bool PyString_Check(IntPtr ob)
...
+                    IntPtr.Zero,
+                    IntPtr.Zero
+                ).DangerousGetAddress();
```
Comment (lostmsu) [inline src/runtime/runtime.cs:1535]: Better use `DangerousMoveToPointer` or `DangerousMoveToPointerOrNull`

**Review by lostmsu — CHANGES_REQUESTED** [review]

**Comment by lostmsu** [comment]
@filmor requested some minor tweaks

File: CHANGELOG.md [hunk 3]
```diff
@@ -71,6 +71,7 @@ One must now either use enum members (e.g. `MyEnum.Option`), or use enum constru
 -   Exception stacktraces on `PythonException.StackTrace` are now properly formatted
 -   Providing an invalid type parameter to a generic type or method produces a helpful Python error
 -   Empty parameter names (as can be generated from F#) do not cause crashes
+-   Unicode strings with surrogates get truncated when converting from Python
```
[the remark this answers is not in the document]
Comment (pkese) [reply CHANGELOG.md]: Thanks, I've applied the suggested change in today's commit.

File: src/runtime/runtime.cs [hunk 4]
```diff
@@ -1646,11 +1619,12 @@ internal static string GetManagedString(IntPtr op)
             if (type == PyUnicodeType)
             {
                 using var p = PyUnicode_AsUTF16String(new BorrowedReference(op));
-                int length = (int)PyUnicode_GetSize(op);
-                char* codePoints = (char*)PyBytes_AsString(p.DangerousGetAddress());
+                var bytesPtr = p.DangerousGetAddress();
+                int bytesLength = (int)Runtime.PyBytes_Size(bytesPtr);
+                char* codePoints = (char*)PyBytes_AsString(bytesPtr);
```
Comment (lostmsu) [inline src/runtime/runtime.cs:1624]: NIT: better add overload for `PyBytes_AsString` that takes `BorrowedReference`. This would remove unnecessary call to `DangerousGetAddress` above.

**Comment by lostmsu** [comment]
@filmor it did not fail on rerun. I think we can merge this in, and investigate if it will reappear.

**Review by lostmsu — APPROVED** [review]
